internal/armada/server.go

```diff
@@ -98,7 +113,9 @@ func Serve(config *configuration.ArmadaConfig, healthChecks *health.MultiChecker
 		jobStatusProcessor := repository.NewEventJobStatusProcessor(config.Events.JobStatusQueue, jobRepository, eventStream, jobStatusBatcher)
 		jobStatusProcessor.Start()
 
-		teardown = func() {
+		// TODO Teardown functions should return an error that can be logged/whatever by the caller.
```

Comment:
I guess the question would be whether the caller could do anything useful with the error.

Reply:
You can wrap errors on the way up, such that when you finally log an error and exit in `main` the logged error allows you to understand the sequence of function calls that caused the error. When calling `panic`, we lose all that information (`panic` is non-local return) and we may not be able to easily understand what sequence of calls were made.

Reply:
yeah I wasn't arguing that it should panic, I was more meaning that this function could either a) return an error if it fails  or b) return void and log any errors that it encounters.  My assumption here is that this is just teardown code whereby if things go wrong, there's not much you can do other than continue on and assume you may have left some resources not cleaned up.  That being the case it *can* be better to choose b), especially if the caller can't do anything useful with the error,  although whether it is in this case her I'm less certain.

Reply:
Ah, I see your point.

Something worth considering is integration with the `errgroups` package. If we run all services within an `ErrGroup` (as opposed to a `WaitGroup`), returning an error can trigger cancelling a context, which in turn can be used to trigger graceful shutdown of other services running in parallel. We do this by giving all services running in parallel a reference to the same context, which then acts as a channel for carrying error information between services. We could do this using both approach a) and b). I'll think about what would be best.
https://pkg.go.dev/golang.org/x/sync/errgroup